src/scalar/mod.rs

```diff
@@ -0,0 +1,135 @@
+use std::{
+    fmt::Display,
+    ops::{Add, AddAssign, Mul, MulAssign, Sub, SubAssign},
+};
+
+use nalgebra::{ClosedAdd, ClosedDiv, ClosedMul, ClosedSub, ComplexField, SimdRealField};
+use num_traits::{Pow, Signed};
+
+use crate::vector::VectorView;
+
+pub trait Scalar:
+    nalgebra::Scalar
+    + From<f64>
+    + Display
+    + SimdRealField
+    + ComplexField
+    + Copy
+    + ClosedSub
+    + From<f64>
+    + ClosedMul
+    + ClosedDiv
+    + ClosedAdd
+    + Signed
+    + Into<f64>
+    + PartialOrd
+    + Pow<Self, Output = Self>
+    + Pow<i32, Output = Self>
+{
+    const EPSILON: Self;
+    const INFINITY: Self;
+}
+
+pub type IndexType = usize;
+
+impl Scalar for f64 {
+    const EPSILON: Self = f64::EPSILON;
+    const INFINITY: Self = f64::INFINITY;
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct Scale<E: Scalar>(pub E);
+
+impl<E: Scalar> Scale<E> {
+    #[inline]
+    pub fn value(self) -> E {
+        self.0
+    }
+}
+
+#[inline]
+pub fn scale<E: Scalar>(value: E) -> Scale<E> {
+    Scale(value)
+}
+
+//TODO: Is it possible for us to need RhsE != LhsE?
+impl<E: Scalar> Mul<Scale<E>> for Scale<E> {
```

Comment:
a lot of this code implementing the ops is duplication, I would create a macro and use it like:
```rust
impl_bin_op!(Add, add, +)
impl_bin_op!(Mul, mul, *)
// ... etc
```

Reply:
Implemented